sql/connect/server/src/main/scala/org/apache/spark/sql/connect/ml/MLHandler.scala

```diff
@@ -0,0 +1,283 @@
+  }
+}
+
+// MLHandler is a utility to group all ML operations
+object MLHandler extends Logging {
+  def handleMlCommand(
+      sessionHolder: SessionHolder,
+      mlCommand: proto.MlCommand): proto.MlCommandResult = {
+
+    val mlCache = sessionHolder.mlCache
+
+    mlCommand.getCommandCase match {
+      case proto.MlCommand.CommandCase.FIT =>
+        val fitCmd = mlCommand.getFit
+        val estimatorProto = fitCmd.getEstimator
+        assert(estimatorProto.getType == proto.MlOperator.OperatorType.ESTIMATOR)
+
+        val dataset = MLUtils.parseRelationProto(fitCmd.getDataset, sessionHolder)
+        val estimator = MLUtils.getEstimator(estimatorProto, Some(fitCmd.getParams))
+        val model = estimator.fit(dataset).asInstanceOf[Model[_]]
+        val id = mlCache.register(model)
+        proto.MlCommandResult
+          .newBuilder()
+          .setOperatorInfo(
+            proto.MlCommandResult.MlOperatorInfo
+              .newBuilder()
+              .setObjRef(proto.ObjectRef.newBuilder().setId(id)))
+          .build()
+
+      case proto.MlCommand.CommandCase.FETCH_ATTR =>
+        val args = mlCommand.getFetchAttr.getArgsList.asScala.toArray
+        val helper = AttributeHelper(
+          sessionHolder,
+          mlCommand.getFetchAttr.getObjRef.getId,
+          Option(mlCommand.getFetchAttr.getMethod),
+          args)
+        val attrResult = helper.getAttribute
+        attrResult match {
+          case s: Summary =>
+            val id = mlCache.register(s)
+            proto.MlCommandResult.newBuilder().setSummary(id).build()
+          case _ =>
+            val param = Serializer.serializeParam(attrResult)
+            proto.MlCommandResult.newBuilder().setParam(param).build()
+        }
+
+      case proto.MlCommand.CommandCase.DELETE =>
+        val modelId = mlCommand.getDelete.getObjRef.getId
+        var result = false
+        if (!modelId.contains(".")) {
+          mlCache.remove(modelId)
+          result = true
+        }
+        proto.MlCommandResult
+          .newBuilder()
+          .setParam(
+            proto.Param
+              .newBuilder()
+              .setLiteral(LiteralValueProtoConverter.toLiteralProto(result))
+              .build())
+          .build()
+
+      case proto.MlCommand.CommandCase.WRITE =>
+        mlCommand.getWrite.getTypeCase match {
+          case proto.MlCommand.Writer.TypeCase.OBJ_REF => // save a model
+            val modelId = mlCommand.getWrite.getObjRef.getId
+            val model = mlCache.get(modelId).asInstanceOf[Model[_]]
+            val copiedModel = model.copy(ParamMap.empty).asInstanceOf[Model[_]]
+            MLUtils.setInstanceParams(copiedModel, mlCommand.getWrite.getParams)
+
+            copiedModel match {
+              case m: MLWritable => MLUtils.write(m, mlCommand.getWrite)
+              case _ => throw new RuntimeException("Failed to handle model.save")
```

Comment:
I think this is my last bigger comment, I would kindly ask to avoid using RuntimeExceptions and instead create a new MLException in SparkConnect that inherits from SparkException and uses the proper message templates.

What you need for that:

- Use XX000 as the error code https://github.com/apache/spark/blob/master/common/utils/src/main/resources/error/error-states.json#L7474
- create a spark connect ML related sub class similar to asof join https://github.com/apache/spark/blob/master/common/utils/src/main/resources/error/error-conditions.json#L85-L107 the error classes must be ordered alphabetically :)
- use the sqlstate from above
- create new sub classes for the error here with proper message templates etc. (cf. https://github.com/apache/spark/blob/13315eeec07e2aebcc05dfee762bbd060ae192ec/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/joinTypes.scala#L148-L153)

Please do this for all exceptions used here.

Reply:
Thx for the comment. Done.